src/uct/tcp/tcp_ep.c

```diff
@@ -465,6 +625,69 @@ ucs_status_t uct_tcp_ep_create(const uct_ep_params_t *params,
     return UCS_OK;
 }
 
+ucs_status_t uct_tcp_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *ep_addr)
+{
+    uct_tcp_ep_t *ep        = ucs_derived_of(tl_ep, uct_tcp_ep_t);
+    uct_tcp_ep_addr_t *addr = (uct_tcp_ep_addr_t*)ep_addr;
+
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
+
+    addr->ptr_map_key = ep->cm_id.ptr_map_key;
+    return uct_iface_get_address(tl_ep->iface,
+                                 (uct_iface_addr_t*)&addr->iface_addr);
+}
+
+ucs_status_t uct_tcp_ep_connect_to_ep(uct_ep_h tl_ep,
+                                      const uct_device_addr_t *dev_addr,
+                                      const uct_ep_addr_t *ep_addr)
+{
+    uct_tcp_ep_t *ep                    = ucs_derived_of(tl_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t UCS_V_UNUSED *iface = ucs_derived_of(ep->super.super.iface,
+                                                         uct_tcp_iface_t);
+    uct_tcp_ep_addr_t *addr             = (uct_tcp_ep_addr_t*)ep_addr;
+    int connect_to_self;
+
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
+
+    if (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) {
+        /* CONN_REQ was already received by the EP, no need for any actions
+         * anymore */
+        ucs_assert(uct_tcp_ep_ptr_map_retrieve(iface,
+                                               ep->cm_id.ptr_map_key) == NULL);
+        return UCS_OK;
+    }
+
+    uct_tcp_ep_set_dest_addr(dev_addr, (uct_iface_addr_t*)&addr->iface_addr,
+                             &ep->peer_addr);
+
+    /* Address has to be set before checking if the EP is connected to self */
+    connect_to_self = uct_tcp_ep_is_self(ep);
+    if (!uct_tcp_cm_ep_accept_conn(ep) || connect_to_self) {
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED);
+
+        if (!connect_to_self) {
+            /* EP that are created as CONNECT_TO_EP has to be full-duplex, set RX
+             * capability as well as TX (that's set in uct_tcp_ep_connect()) */
+            uct_tcp_ep_add_ctx_cap(ep, UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+        }
+
+        uct_tcp_ep_ptr_map_del(ep);
+
+        /* Use remote peer connection sequence number value, since the EP has to
+         * send the CONN_REQ to the peer has to find its EP in the EP PTR map */
+        ep->cm_id.ptr_map_key = addr->ptr_map_key;
+        return uct_tcp_ep_connect(ep);
+    }
+
+    if (ep->conn_state != UCT_TCP_EP_CONN_STATE_CONNECTED) {
```

Comment:
why need this if?
i would expect that the code which handles incoming conn_req would set these flags